hazelcast/src/test/java/com/hazelcast/instance/DefaultAddressPickerInterfacesTest.java

```diff
@@ -0,0 +1,429 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.instance;
+
+import static com.hazelcast.instance.DefaultAddressPickerInterfacesTest.NetworkInterfaceOptions.builder;
+import static com.hazelcast.instance.TestUtil.setSystemProperty;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static org.hamcrest.CoreMatchers.anyOf;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.spi.properties.HazelcastProperties;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+
+/**
+ * This class contains PowerMock driven tests which emulate different NetworkInterfaces configurations. The tests check if the
+ * DefaultAdddressPicker chooses an expected bind address.
+ * <p>
+ * Given: Default Hazelcast Config is used and no Interface definition network configuration is set. No system property which
+ * could influence address picking is used.
+ * </p>
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({ DefaultAddressPicker.class })
+@Category({ QuickTest.class, ParallelTest.class })
+public class DefaultAddressPickerInterfacesTest {
+
+    private static final String SYS_PROP_PREFER_IPV4 = "java.net.preferIPv4Stack";
+    private static String origPreferIpv4;
+    private static String origPreferIpv4Hazelcast;
+
+    private final ILogger logger = Logger.getLogger(AddressPicker.class);
+    private final Config config = new Config();
+    private final HazelcastProperties hazelcastProperties = new HazelcastProperties(config);
+
+    /**
+     * Test class setup: Prefer IPv4, but only by standard Java property.
+     */
+    @BeforeClass
+    public static void beforeClass() {
+        origPreferIpv4 = setSystemProperty(SYS_PROP_PREFER_IPV4, "true");
+        origPreferIpv4Hazelcast = setSystemProperty("hazelcast.prefer.ipv4.stack", "false");
```

Comment:
You can use `GroupProperty.PREFER_IPv4_STACK.getName()` instead of `hazelcast.prefer.ipv4.stack` here

Reply:
I tend to use the string directly in the test. It can catch backward compatibility issues when the GroupProperty is refactored.

Reply:
Right, seems reasonable.

Reply:
there is convenience for this. see it in [action](https://github.com/jerrinot/hazelcast/blob/0862c483877c7ba750cda91700b3d4e47bf941b9/hazelcast/src/test/java/com/hazelcast/spi/discovery/multicast/MemberToMemberDiscoveryTest.java) It does auto-cleaning and hopefully makes the test somewhat cleaner.

Reply:
:+1: will fix